File: api/src/test/scala/com/github/pflooky/datacaterer/api/DataCatererConfigurationBuilderTest.scala
package com.github.pflooky.datacaterer.api

import com.github.pflooky.datacaterer.api.model.Constants.{DEFAULT_CASSANDRA_PASSWORD, DEFAULT_CASSANDRA_USERNAME, DEFAULT_KAFKA_URL, DEFAULT_MYSQL_PASSWORD, DEFAULT_MYSQL_URL, DEFAULT_MYSQL_USERNAME, DEFAULT_POSTGRES_PASSWORD, DEFAULT_POSTGRES_URL, DEFAULT_POSTGRES_USERNAME, DEFAULT_SOLACE_CONNECTION_FACTORY, DEFAULT_SOLACE_INITIAL_CONTEXT_FACTORY, DEFAULT_SOLACE_PASSWORD, DEFAULT_SOLACE_URL, DEFAULT_SOLACE_USERNAME, DEFAULT_SOLACE_VPN_NAME}
import com.github.pflooky.datacaterer.api.model.{FlagsConfig, FoldersConfig, GenerationConfig, MetadataConfig}
import org.junit.runner.RunWith
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class DataCatererConfigurationBuilderTest extends AnyFunSuite {

  test("Can create basic configuration with defaults") {
    val result = DataCatererConfigurationBuilder().build

    assert(result.flagsConfig == FlagsConfig())
    assert(result.foldersConfig == FoldersConfig())
    assert(result.metadataConfig == MetadataConfig())
    assert(result.generationConfig == GenerationConfig())
    assert(result.connectionConfigByName.isEmpty)
    assert(result.runtimeConfig.size == 11)
    assert(result.master == "local[*]")
  }

  test("Can create postgres connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .postgres("my_postgres")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_postgres"))
    val config = result("my_postgres")
    assert(config("url") == DEFAULT_POSTGRES_URL)
    assert(config("user") == DEFAULT_POSTGRES_USERNAME)
    assert(config("password") == DEFAULT_POSTGRES_PASSWORD)
    assert(config("format") == "jdbc")
    assert(config("driver") == "org.postgresql.Driver")
  }

  test("Can create postgres connection with custom configuration") {
    val result = DataCatererConfigurationBuilder()
      .postgres("my_postgres", "jdbc:postgresql://localhost:5432/customer", options = Map("stringtype" -> "undefined"))
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_postgres"))
    val config = result("my_postgres")
    assert(config.size == 6)
    assert(config("url") == "jdbc:postgresql://localhost:5432/customer")
    assert(config("stringtype") == "undefined")
  }

  test("Can create mysql connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .mysql("my_mysql")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_mysql"))
    val config = result("my_mysql")
    assert(config("url") == DEFAULT_MYSQL_URL)
    assert(config("user") == DEFAULT_MYSQL_USERNAME)
    assert(config("password") == DEFAULT_MYSQL_PASSWORD)
    assert(config("format") == "jdbc")
    assert(config("driver") == "com.mysql.cj.jdbc.Driver")
  }

  test("Can create cassandra connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .cassandra("my_cassandra")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_cassandra"))
    val config = result("my_cassandra")
    assert(config("spark.cassandra.connection.host") == "cassandraserver")
    assert(config("spark.cassandra.connection.port") == "9042")
    assert(config("spark.cassandra.auth.username") == DEFAULT_CASSANDRA_USERNAME)
    assert(config("spark.cassandra.auth.password") == DEFAULT_CASSANDRA_PASSWORD)
    assert(config("format") == "org.apache.spark.sql.cassandra")
  }

  test("Can create solace connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .solace("my_solace")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_solace"))
    val config = result("my_solace")
    assert(config("url") == DEFAULT_SOLACE_URL)
    assert(config("user") == DEFAULT_SOLACE_USERNAME)
    assert(config("password") == DEFAULT_SOLACE_PASSWORD)
    assert(config("format") == "jms")
    assert(config("vpnName") == DEFAULT_SOLACE_VPN_NAME)
    assert(config("connectionFactory") == DEFAULT_SOLACE_CONNECTION_FACTORY)
    assert(config("initialContextFactory") == DEFAULT_SOLACE_INITIAL_CONTEXT_FACTORY)
  }

  test("Can create kafka connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .kafka("my_kafka")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_kafka"))
    val config = result("my_kafka")
    assert(config("kafka.bootstrap.servers") == DEFAULT_KAFKA_URL)
    assert(config("format") == "kafka")
  }

  test("Can create http connection configuration") {
    val result = DataCatererConfigurationBuilder()
      .http("my_http", "user", "pw")
      .build
      .connectionConfigByName

    assert(result.size == 1)
    assert(result.contains("my_http"))
    val config = result("my_http")
    assert(config("user") == "user")
    assert(config("password") == "pw")
  }

  test("Can enable/disable flags") {
    val result = DataCatererConfigurationBuilder()
      .enableCount(false)
      .enableGenerateData(false)
      .enableDeleteGeneratedRecords(true)
      .enableGeneratePlanAndTasks(true)
      .enableUniqueCheck(true)
      .enableFailOnError(false)
      .enableRecordTracking(true)
      .enableSaveReports(true)
      .enableSinkMetadata(true)
      .enableValidation(true)
      .build
      .flagsConfig

    assert(!result.enableCount)
    assert(!result.enableGenerateData)
    assert(result.enableDeleteGeneratedRecords)
    assert(result.enableGeneratePlanAndTasks)
    assert(result.enableUniqueCheck)
    assert(!result.enableFailOnError)
    assert(result.enableRecordTracking)
    assert(result.enableSaveReports)
    assert(result.enableSinkMetadata)
    assert(result.enableValidation)
  }

  test("Can alter folder paths") {
    val result = DataCatererConfigurationBuilder()
      .planFilePath("/my_plan")
      .taskFolderPath("/my_task")
      .recordTrackingFolderPath("/my_record_tracking")
      .validationFolderPath("/my_validation")
      .generatedReportsFolderPath("/my_generation_results")
      .generatedPlanAndTaskFolderPath("/my_generated_plan_tasks")
      .build
      .foldersConfig

    assert(result.planFilePath == "/my_plan")
    assert(result.taskFolderPath == "/my_task")
    assert(result.recordTrackingFolderPath == "/my_record_tracking")
    assert(result.validationFolderPath == "/my_validation")
    assert(result.generatedReportsFolderPath == "/my_generation_results")
    assert(result.generatedPlanAndTaskFolderPath == "/my_generated_plan_tasks")
  }

  test("Can alter metadata configurations") {
    val result = DataCatererConfigurationBuilder()
      .numRecordsFromDataSourceForDataProfiling(1)
      .numRecordsForAnalysisForDataProfiling(2)
      .numGeneratedSamples(3)
      .oneOfMinCount(100)
      .oneOfDistinctCountVsCountThreshold(0.3)
      .build
      .metadataConfig

    assert(result.numRecordsFromDataSource == 1)
    assert(result.numRecordsForAnalysis == 2)
    assert(result.numGeneratedSamples == 3)
    assert(result.oneOfMinCount == 100)
    assert(result.oneOfDistinctCountVsCountThreshold == 0.3)
  }

  test("Can alter generation configurations") {
    val result = DataCatererConfigurationBuilder()
      .numRecordsPerBatch(100)
      .numRecordsPerStep(10)
      .build
      .generationConfig

    assert(result.numRecordsPerBatch == 100)
    assert(result.numRecordsPerStep.contains(10))
  }
}

File: api-java/src/test/java/com/github/pflooky/datacaterer/api/java/model/config/DataCatererConfigurationBuilderTest.java
package com.github.pflooky.datacaterer.api.java.model.config;

import com.github.pflooky.datacaterer.api.model.Constants;
import org.junit.Test;

import java.util.Collections;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class DataCatererConfigurationBuilderTest {

    @Test
    public void canCreateDataCatererConfigurationWithConnections() {
        var result = new DataCatererConfigurationBuilder()
                .master("local[1]")
                .runtimeConfig(Map.of("spark.conf", "my_value"))
                .addRuntimeConfig("spark.new.conf", "second_val")
                .csv("my_csv", "/path/csv")
                .parquet("my_parquet", "/path/parquet")
                .orc("my_orc", "/path/orc")
                .json("my_json", "/path/json")
                .postgres("my_postgres", "jdbc:postgresql://localhost:5432")
                .mysql("my_mysql", "jdbc:mysql://localhost:3306")
                .cassandra("my_cassandra", "localhost:9042")
                .jms("my_jms", "localhost:1234", "admin", "admin", Collections.emptyMap())
                .solace("my_solace", "smf://localhost:55554")
                .kafka("my_kafka", "localhost:9092")
                .http("my_http")
                .configuration().build();

        assertEquals(11, result.connectionConfigByName().size());
        assertTrue(result.connectionConfigByName().contains("my_csv"));
        assertTrue(result.connectionConfigByName().get("my_csv").get().get(Constants.FORMAT()).contains(Constants.CSV()));
        assertTrue(result.connectionConfigByName().get("my_csv").get().get(Constants.PATH()).contains("/path/csv"));
        assertTrue(result.connectionConfigByName().contains("my_parquet"));
        assertTrue(result.connectionConfigByName().get("my_parquet").get().get(Constants.FORMAT()).contains(Constants.PARQUET()));
        assertTrue(result.connectionConfigByName().get("my_parquet").get().get(Constants.PATH()).contains("/path/parquet"));
        assertTrue(result.connectionConfigByName().contains("my_orc"));
        assertTrue(result.connectionConfigByName().get("my_orc").get().get(Constants.FORMAT()).contains(Constants.ORC()));
        assertTrue(result.connectionConfigByName().get("my_orc").get().get(Constants.PATH()).contains("/path/orc"));
        assertTrue(result.connectionConfigByName().contains("my_json"));
        assertTrue(result.connectionConfigByName().get("my_json").get().get(Constants.FORMAT()).contains(Constants.JSON()));
        assertTrue(result.connectionConfigByName().get("my_json").get().get(Constants.PATH()).contains("/path/json"));

        assertTrue(result.connectionConfigByName().contains("my_postgres"));
        var postgres = result.connectionConfigByName().get("my_postgres").get();
        assertTrue(postgres.get(Constants.FORMAT()).contains(Constants.JDBC()));
        assertTrue(postgres.get(Constants.URL()).contains("jdbc:postgresql://localhost:5432"));
        assertTrue(postgres.get(Constants.USERNAME()).contains("postgres"));
        assertTrue(postgres.get(Constants.PASSWORD()).contains("postgres"));
        assertTrue(postgres.get(Constants.DRIVER()).contains(Constants.POSTGRES_DRIVER()));

        assertTrue(result.connectionConfigByName().contains("my_mysql"));
        var mysql = result.connectionConfigByName().get("my_mysql").get();
        assertTrue(mysql.get(Constants.FORMAT()).contains(Constants.JDBC()));
        assertTrue(mysql.get(Constants.URL()).contains("jdbc:mysql://localhost:3306"));
        assertTrue(mysql.get(Constants.USERNAME()).contains("root"));
        assertTrue(mysql.get(Constants.PASSWORD()).contains("root"));
        assertTrue(mysql.get(Constants.DRIVER()).contains(Constants.MYSQL_DRIVER()));

        assertTrue(result.connectionConfigByName().contains("my_cassandra"));
        var cassandra = result.connectionConfigByName().get("my_cassandra").get();
        assertTrue(cassandra.get(Constants.FORMAT()).contains(Constants.CASSANDRA()));
        assertTrue(cassandra.get("spark.cassandra.connection.host").contains("localhost"));
        assertTrue(cassandra.get("spark.cassandra.connection.port").contains("9042"));
        assertTrue(cassandra.get("spark.cassandra.auth.username").contains("cassandra"));
        assertTrue(cassandra.get("spark.cassandra.auth.password").contains("cassandra"));

        assertTrue(result.connectionConfigByName().contains("my_jms"));
        assertTrue(result.connectionConfigByName().get("my_jms").get().get(Constants.FORMAT()).contains(Constants.JMS()));
        assertTrue(result.connectionConfigByName().contains("my_solace"));
        var solace = result.connectionConfigByName().get("my_solace").get();
        assertTrue(solace.get(Constants.FORMAT()).contains(Constants.JMS()));
        assertTrue(solace.get(Constants.URL()).contains("smf://localhost:55554"));
        assertTrue(solace.get(Constants.USERNAME()).contains("admin"));
        assertTrue(solace.get(Constants.PASSWORD()).contains("admin"));
        assertTrue(solace.get(Constants.JMS_VPN_NAME()).contains("default"));
        assertTrue(solace.get(Constants.JMS_CONNECTION_FACTORY()).contains(Constants.DEFAULT_SOLACE_CONNECTION_FACTORY()));
        assertTrue(solace.get(Constants.JMS_INITIAL_CONTEXT_FACTORY()).contains(Constants.DEFAULT_SOLACE_INITIAL_CONTEXT_FACTORY()));

        assertTrue(result.connectionConfigByName().contains("my_kafka"));
        assertTrue(result.connectionConfigByName().get("my_kafka").get().get(Constants.FORMAT()).contains(Constants.KAFKA()));
        assertTrue(result.connectionConfigByName().get("my_kafka").get().get("kafka.bootstrap.servers").contains("localhost:9092"));

        assertTrue(result.connectionConfigByName().contains("my_http"));
        assertTrue(result.connectionConfigByName().get("my_http").get().get(Constants.FORMAT()).contains(Constants.HTTP()));
    }

    @Test
    public void canCreateDataCatererConfigurationWithFlags() {
        var result = new DataCatererConfigurationBuilder()
                .enableCount(false)
                .enableGenerateData(false)
                .enableDeleteGeneratedRecords(true)
                .enableFailOnError(false)
                .enableGeneratePlanAndTasks(true)
                .enableRecordTracking(true)
                .enableSaveReports(false)
                .enableSinkMetadata(true)
                .enableUniqueCheck(true)
                .enableValidation(true)
                .configuration().build();

        assertFalse(result.flagsConfig().enableCount());
        assertFalse(result.flagsConfig().enableGenerateData());
        assertTrue(result.flagsConfig().enableDeleteGeneratedRecords());
        assertFalse(result.flagsConfig().enableFailOnError());
        assertTrue(result.flagsConfig().enableGeneratePlanAndTasks());
        assertTrue(result.flagsConfig().enableRecordTracking());
        assertFalse(result.flagsConfig().enableSaveReports());
        assertTrue(result.flagsConfig().enableSinkMetadata());
        assertTrue(result.flagsConfig().enableUniqueCheck());
        assertTrue(result.flagsConfig().enableValidation());
    }

    @Test
    public void canCreateDataCatererConfigurationWithFolders() {
        var result = new DataCatererConfigurationBuilder()
                .planFilePath("/my/plan")
                .taskFolderPath("/my/task")
                .recordTrackingFolderPath("/my/record")
                .validationFolderPath("/my/validation")
                .generatedReportsFolderPath("/my/report")
                .generatedPlanAndTaskFolderPath("/my/gen")
                .configuration().build();

        assertEquals("/my/plan", result.foldersConfig().planFilePath());
        assertEquals("/my/task", result.foldersConfig().taskFolderPath());
        assertEquals("/my/record", result.foldersConfig().recordTrackingFolderPath());
        assertEquals("/my/validation", result.foldersConfig().validationFolderPath());
        assertEquals("/my/report", result.foldersConfig().generatedReportsFolderPath());
        assertEquals("/my/gen", result.foldersConfig().generatedPlanAndTaskFolderPath());
    }

    @Test
    public void canCreateDataCatererConfigurationWithMetadata() {
        var result = new DataCatererConfigurationBuilder()
                .numRecordsFromDataSourceForDataProfiling(10)
                .numRecordsForAnalysisForDataProfiling(10)
                .numGeneratedSamples(1)
                .oneOfMinCount(10)
                .oneOfDistinctCountVsCountThreshold(0.5)
                .configuration().build();

        assertEquals(10, result.metadataConfig().numRecordsFromDataSource());
        assertEquals(10, result.metadataConfig().numRecordsForAnalysis());
        assertEquals(1, result.metadataConfig().numGeneratedSamples());
        assertEquals(10, result.metadataConfig().oneOfMinCount());
        assertEquals(0.5, result.metadataConfig().oneOfDistinctCountVsCountThreshold(), 0.01);
    }

    @Test
    public void canCreateDataCatererConfigurationWithGeneration() {
        var result = new DataCatererConfigurationBuilder()
                .numRecordsPerBatch(10)
                .numRecordsPerStep(10)
                .configuration().build();

        assertEquals(10, result.generationConfig().numRecordsPerBatch());
        assertTrue(result.generationConfig().numRecordsPerStep().contains(10));
    }
}